cdc/sorter/leveldb/leveldb.go

```diff
@@ -0,0 +1,219 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leveldb
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/ticdc/cdc/sorter/leveldb/message"
+	"github.com/pingcap/ticdc/pkg/actor"
+	actormsg "github.com/pingcap/ticdc/pkg/actor/message"
+	"github.com/pingcap/ticdc/pkg/config"
+	"github.com/pingcap/ticdc/pkg/retry"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+	"github.com/syndtr/goleveldb/leveldb/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/semaphore"
+)
+
+// OpenDB opens a leveldb.
+func OpenDB(ctx context.Context, id int, cfg *config.SorterConfig) (*leveldb.DB, error) {
+	lcfg := cfg.LevelDB
+	var option opt.Options
+	option.OpenFilesCacheCapacity = lcfg.MaxOpenFiles / lcfg.Count
+	option.BlockCacheCapacity = lcfg.BlockCacheSize / lcfg.Count
+	option.BlockSize = lcfg.BlockSize
+	option.WriteBuffer = lcfg.WriterBufferSize
+	option.Compression = opt.NoCompression
+	if lcfg.Compression == "snappy" {
+		option.Compression = opt.SnappyCompression
+	}
+	option.CompactionTableSize = lcfg.TargetFileSizeBase
+	option.CompactionL0Trigger = lcfg.CompactionL0Trigger
+	option.WriteL0SlowdownTrigger = lcfg.WriteL0SlowdownTrigger
+	option.WriteL0PauseTrigger = lcfg.WriteL0PauseTrigger
+	option.ErrorIfExist = true
+	option.NoSync = true
+
+	// TODO make sure sorter dir is under data dir.
+	dbDir := filepath.Join(cfg.SortDir, fmt.Sprintf("%04d", id))
+	err := retry.Do(ctx, func() error {
+		err1 := os.RemoveAll(dbDir)
+		if err1 != nil {
+			log.Warn("clean data dir fails",
+				zap.String("dir", dbDir), zap.Error(err1))
+		}
+		return err1
+	},
+		retry.WithBackoffBaseDelay(500), // 0.5s
+		retry.WithMaxTries(4))           // 2s in total.
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	db, err := leveldb.OpenFile(dbDir, &option)
+	return db, errors.Trace(err)
+}
+
+// LevelActor is a leveldb actor, it reads, writes and deletes key value pair
+// in its leveldb.
+type LevelActor struct {
+	id       actor.ID
+	db       *leveldb.DB
+	wb       *leveldb.Batch
+	wbSize   int
+	iterSema *semaphore.Weighted
+	closedWg *sync.WaitGroup
+
+	metricWriteDuration prometheus.Observer
+	metricWriteBytes    prometheus.Observer
+}
+
+var _ actor.Actor = (*LevelActor)(nil)
+
+// NewLevelDBActor returns a leveldb actor.
+func NewLevelDBActor(
+	ctx context.Context, id int, db *leveldb.DB, cfg *config.SorterConfig,
+	wg *sync.WaitGroup, captureAddr string,
+) (*LevelActor, actor.Mailbox, error) {
+	idTag := fmt.Sprint(id)
+	// Write batch size should be larger than block size to save CPU.
+	const batchSizeFactor = 16
```

Comment:
make it a global constant?

Reply:
No, I prefer the minimum visibility scope.

Reply:
I don't think it is a local variable, since block size is a global config, why its factor be hidden in internal logic.

Reply:
But it is not some critical thing. never mind.